### PR TITLE
feat: support unnamed parameters

### PR DESCRIPTION
**Pull request by sakthivelmanii**
Currently Spanner Client Library doesn't support passing positional (or unnamed) parameters while creating the statement. This feature provides the support for unnamed parameters which helps customers to create SQL Statement with unnamed parameter(?)

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-spanner/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClient.java [hunk 1]
```diff
@@ -652,6 +653,14 @@ public Dialect getDialect() {
     }
   }
 
+  public Future<Dialect> getDialectAsync() {
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClient.java:656]: Could this be package-private?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    return Value.untyped(
+        com.google.protobuf.Value.newBuilder().setStringValue(String.valueOf(value)).build());
+  }
+
+  @SuppressWarnings("unchecked")
+  static <T, U> Iterable<U> convertToTypedIterable(
+      Function<T, U> func, T val, Iterator<?> iterator) {
+    List<U> values = new ArrayList<>();
+    values.add(func.apply(val));
+    iterator.forEachRemaining(value -> values.add(func.apply((T) value)));
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:74]: Why do we need to do this in two separate steps (one for the first item, and then for each remaining item)?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        com.google.protobuf.Value.newBuilder()
+            .setListValue(ListValue.newBuilder().addAllValues(values).build())
+            .build());
+  }
+
+  static <T extends TemporalAccessor> String convertToISO8601(T dateTime) {
+    return ISO_8601_DATE_FORMATTER.format(dateTime);
+  }
+
+  static <T> Value createUntypedValue(T value) {
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:64]: This should only be used for values that should be encoded as a string value.

```suggestion
  static <T> Value createUntypedStringValue(T value) {
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  @SuppressWarnings("unchecked")
+  static <T, U> Iterable<U> convertToTypedIterable(
+      Function<T, U> func, T val, Iterator<?> iterator) {
+    List<U> values = new ArrayList<>();
+    values.add(func.apply(val));
+    iterator.forEachRemaining(value -> values.add(func.apply((T) value)));
+    return values;
+  }
+
+  static <T> Iterable<T> convertToTypedIterable(T val, Iterator<?> iterator) {
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:78]: Could this be:

```suggestion
  static <T> Iterable<T> convertToTypedIterable(T val, Iterator<T> iterator) {
```

Or `Iterator<? extends T>`?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    return convertToTypedIterable(v -> v, val, iterator);
+  }
+
+  static Date convertUtilDateToSpannerDate(java.util.Date date) {
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:82]: There is already a `fromJavaUtilDate` method in `com.google.cloud.Date`. We should (probably) use that one.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    return Date.parseDate(SIMPLE_DATE_FORMATTER.format(date));
+  }
+
+  static Date convertLocalDateToSpannerDate(LocalDate date) {
+    return Date.parseDate(DATE_FORMATTER.format(date));
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:87]: It's more efficient to just get the year, month, and day from the `LocalDate` and construct the `com.google.cloud.Date` from that.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            .setListValue(
+                ListValue.newBuilder()
+                    .addAllValues(
+                        SpannerTypeConverter.convertToTypedIterable(
+                            (val) ->
+                                com.google.protobuf.Value.newBuilder()
+                                    .setStringValue(func.apply(val))
+                                    .build(),
+                            value,
+                            iterator)))
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:103]: I think that this block could be optimized a bit by doing a direct conversion from the iterator straight into a ListValueBuilder, similar to how we are doing that here: https://github.com/googleapis/java-spanner/blob/92494d080c8c6370d28063751cd8aeb22890903a/google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/RandomResultSetGenerator.java#L231

The current implementation first creates a new ArrayList from the Iterator, and then adds the ArrayList to the ListValueBuilder.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            .build());
+  }
+
+  static ZonedDateTime convertToUTCTimezone(LocalDateTime localDateTime) {
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:107]: Can we rename these `convert...` methods into `atUTC` or similar. We are not 'converting' the values. We are just saying 'show the value in UTC'.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 9]
```diff
@@ -246,4 +248,56 @@ StringBuilder toString(StringBuilder b) {
     }
     return b;
   }
+
+  public static final class StatementFactory {
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:252]: nit: All public classes should have Javadoc

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 10]
```diff
@@ -246,4 +248,56 @@ StringBuilder toString(StringBuilder b) {
...
+     *     automatically
+     *     <ul>
+     *       <li>LocalDateTime
+     *       <li>OffsetDateTime
+     *       <li>ZonedDateTime
+     *     </ul>
+     *     <p>
+     * @see DatabaseClient#newStatementFactory
+     */
+    public Statement of(String sql, Object... values) {
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:286]: I think that it would be good to clearly call out what the purpose is of this method, also in the method name.

```suggestion
    public Statement withUnnamedParameters(String sql, Object... values) {
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 11]
```diff
@@ -246,4 +248,56 @@ StringBuilder toString(StringBuilder b) {
...
+    StatementFactory(Dialect dialect) {
+      this.dialect = dialect;
+    }
+
+    public Statement of(String sql) {
+      return Statement.of(sql);
+    }
+
+    /**
+     * @param sql SQL statement with unnamed parameter denoted as ?
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:264]:
```suggestion
     * @param sql SQL statement with unnamed parameters denoted as ?
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 12]
```diff
@@ -246,4 +248,56 @@ StringBuilder toString(StringBuilder b) {
...
+     * @param values list of values which needs to replace ? in the sql
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:265]:
```suggestion
     * @param values positional list of values for the unnamed parameters in the SQL string
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 13]
```diff
@@ -246,4 +248,56 @@ StringBuilder toString(StringBuilder b) {
...
+     * @return Statement object
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:266]: Either remove the `@return` tag or add a description of what is returned, e.g.:
```suggestion
     * @return Statement a statement that can be executed on Spanner 
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    return localDateTime.atZone(UTC_ZONE);
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:108]: Would it not make more sense to interpret this value with the system default timezone?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 15]
```diff
@@ -246,4 +248,56 @@ StringBuilder toString(StringBuilder b) {
...
+     *     </ul>
+     *     <p>For Timestamp column, following types are supported. All the dates should be in UTC
+     *     format. Incase if the timezone is not in UTC, spanner client will convert that to UTC
+     *     automatically
+     *     <ul>
+     *       <li>LocalDateTime
+     *       <li>OffsetDateTime
+     *       <li>ZonedDateTime
+     *     </ul>
+     *     <p>
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:283]: It is customary to add this kind of description at the start of the Javadoc, before the parameters and return type tags, so this block should be moved up.

```suggestion
     *     <p>This function accepts a SQL statement with unnamed parameters (?), and accepts a
     *     positional list of objects as the values of the unnamed parameters. Primitive types are supported.
     *     <p>For Date values, the following types are supported:
     *     <ul>
     *       <li>java.util.Date
     *       <li>java.time.LocalDate
     *       <li>com.google.cloud.Date
     *     </ul>
     *     <p>For Timestamp values, the following types are supported:
     *     <ul>
     *       <li>LocalDateTime
     *       <li>OffsetDateTime
     *       <li>ZonedDateTime
     *     </ul>
     *     <p>
```

We also need a comment on how `LocalDateTime` is interpreted. The current implementation interprets it in UTC, but I'm not sure that is the most logical. `LocalDateTime` is intended to be interpreted using the local timezone, so we should consider that here as well.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java [hunk 16]
```diff
@@ -824,6 +838,166 @@ public static Value structArray(Type elementType, @Nullable Iterable<Struct> v)
...
+    }
+    if (value instanceof byte[]) {
+      return Value.bytes(ByteArray.copyFrom((byte[]) value));
+    }
+    if (value instanceof Date) {
+      return Value.date((Date) value);
+    }
+    if (value instanceof java.util.Date) {
+      return Value.date(SpannerTypeConverter.convertUtilDateToSpannerDate((java.util.Date) value));
+    }
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:871]: I'm not entirely sure about this conversion. `java.util.Date` is actually more like a `TIMESTAMP` than a `DATE`. We should consider:
1. Just disallowing the use of `java.util.Date` in the first place, as this class has serious flaws.
2. Or converting it to a `TIMESTAMP` rather than a `DATE`.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java [hunk 17]
```diff
@@ -824,6 +838,166 @@ public static Value structArray(Type elementType, @Nullable Iterable<Struct> v)
...
+    if (value instanceof Iterable<?>) {
+      Iterator<?> iterator = ((Iterable<?>) value).iterator();
+      if (!iterator.hasNext()) {
+        return createUntypedArrayValue(Stream.empty());
+      }
+      Object object = iterator.next();
+      if (object instanceof Boolean) {
+        return Value.boolArray(convertToTypedIterable((Boolean) object, iterator));
+      }
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:907]: Is there any way that we can make this if block just depend on the conversions above, instead of repeating the same type of conversion here? (Meaning: Somehow recursively call back into `toValue(Object)` for each value in the iterable, instead of copy-pasting the same conversion in this if block)

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java [hunk 18]
```diff
@@ -4896,6 +4896,23 @@ public void testMetadataUnknownTypes() {
...
 
+  @Test
+  public void testStatementWithIUnnamedParametersParameter() {
+    DatabaseClient client =
+        spanner.getDatabaseClient(DatabaseId.of(TEST_PROJECT, TEST_INSTANCE, TEST_DATABASE));
+
+    Statement statement = client.newStatementFactory().of("select id from test where b=?", true);
+    Statement generatedStatement =
+        Statement.newBuilder("select id from test where b=@p1").bind("p1").to(true).build();
+    mockSpanner.putStatementResult(StatementResult.query(generatedStatement, SELECT1_RESULTSET));
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java:4907]: We should add tests for statements with:
1. Comments with question marks inside the comments
2. String literals with question marks inside the string literals
3. Statement hints at the start

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java (hunk 4 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:78]: It's not possible because we can check `instanceof`  only with `Iterable<?>`. So We are converting the object to `Iterable<?>`.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClient.java [hunk 19]
```diff
@@ -606,4 +607,16 @@ ServerStream<BatchWriteResponse> batchWriteAtLeastOnce(
...
    */
   long executePartitionedUpdate(Statement stmt, UpdateOption... options);
+
+  /**
+   * Returns StatementFactory for the given dialect. With StatementFactory, unnamed parameterized
+   * queries can be passed along with the values to create a Statement.
+   *
+   * <p>Examples using {@link StatementFactory}
+   *
+   * <p>databaseClient.newStatementFactory().of("SELECT NAME FROM TABLE WHERE ID = ?", 10)
```
[the remark this answers is not in the document]
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClient.java]: done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClient.java [hunk 20]
```diff
@@ -606,4 +607,16 @@ ServerStream<BatchWriteResponse> batchWriteAtLeastOnce(
...
+   */
+  default StatementFactory newStatementFactory() {
```
[the remark this answers is not in the document]
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClient.java]: done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 21]
```diff
@@ -139,6 +146,21 @@ public Dialect getDialect() {
     return pool.getDialect();
   }
 
+  @Override
+  public StatementFactory newStatementFactory() {
+    if (statementFactory == null) {
+      try {
+        Dialect dialect = getDialectAsync().get(5, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 22]
```diff
@@ -139,6 +146,21 @@ public Dialect getDialect() {
     return pool.getDialect();
   }
 
+  @Override
+  public StatementFactory newStatementFactory() {
+    if (statementFactory == null) {
```
[the remark this answers is not in the document]
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClient.java (hunk 1 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClient.java]: yes. done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java (hunk 5 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java]: done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java (hunk 6 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java]: done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java (hunk 16 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java]: Removed java.util.Date

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java (hunk 17 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:906]: as we discussed, we can't do that because we might have to do a type check again and again for each element of the iterable

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java (hunk 15 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 23]
```diff
@@ -41,6 +46,8 @@ class DatabaseClientImpl implements DatabaseClient {
   @VisibleForTesting final boolean useMultiplexedSessionPartitionedOps;
   @VisibleForTesting final boolean useMultiplexedSessionForRW;
 
+  private StatementFactory statementFactory = null;
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java:49]: nit: this can be removed

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java (hunk 4 above)
Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java:73]: ack

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java (hunk 17 above)
Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:908]: ack

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java (hunk 7 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java (hunk 8 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java (hunk 11 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java (hunk 12 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java (hunk 13 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java]: Done

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java (hunk 18 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java (hunk 14 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerTypeConverter.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java (hunk 9 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:280]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java (hunk 23 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClient.java [hunk 24]
```diff
@@ -606,4 +607,17 @@ ServerStream<BatchWriteResponse> batchWriteAtLeastOnce(
...
+
+  /**
+   * Returns StatementFactory for the given dialect.
+   *
+   * <p>A {@link StatementFactory}, can be used to create statements with unnamed parameters.
+   *
+   * <p>Examples using {@link StatementFactory}
+   *
+   * <p>databaseClient.getStatementFactory().of("SELECT NAME FROM TABLE WHERE ID = ?", 10)
+   */
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClient.java:619]:
```suggestion
  /**
   * Returns a {@link StatementFactory} for the given dialect.
   *
   * <p>A {@link StatementFactory} can be used to create statements with unnamed parameters.
   * This is primarily intended for framework developers who want to integrate the Spanner client
   * with frameworks that use unnamed parameters. Developers who just want to use the Spanner
   * client in their application, should use named parameters.
   *
   * <p>Examples using {@link StatementFactory}
   *
   * <pre>{@code
   * Statement statement = databaseClient
   *     .getStatementFactory()
   *     .withUnnamedParameters("SELECT NAME FROM TABLE WHERE ID = ?", 10);
   * }</pre>
   */
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 25]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
     }
     return b;
   }
+
+  /**
+   * Factory for creating {@link Statement}.
+   *
+   * <p>This factory class supports creating {@link Statement} with positional(or unnamed)
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:255]: nit: 
```suggestion
   * <p>This factory class supports creating {@link Statement} with positional (or unnamed)
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 26]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
...
+   * <h2>Usage Example</h2>
+   *
+   * Simple SQL query
+   *
+   * <pre>{@code
+   * Statement statement = databaseClient.getStatementFactory()
+   *     .withUnnamedParameters("SELECT * FROM TABLE WHERE ID = ?", 10L)
+   * }</pre>
+   *
+   * How to use SQL queries with IN command
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:269]:
```suggestion
   * How to use arrays with the IN operator
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 27]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
...
+   *
+   * <h2>Usage Example</h2>
+   *
+   * Simple SQL query
+   *
+   * <pre>{@code
+   * Statement statement = databaseClient.getStatementFactory()
+   *     .withUnnamedParameters("SELECT * FROM TABLE WHERE ID = ?", 10L)
+   * }</pre>
+   *
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:268]: Can we add an example here (so after the simple query, but before the array example) that shows how to use it with multiple parameters. This could for example be an INSERT statement that inserts a row with 2 or 3 column values.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 28]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
     }
     return b;
   }
+
+  /**
+   * Factory for creating {@link Statement}.
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:253]:
```suggestion
   * Factory for creating {@link Statement}s with unnamed parameters.
   * This class is primarily intended for framework developers who want to integrate the Spanner
   * client with a framework that uses unnamed parameters. Developers who want to use the
   * Spanner client in their application, should use named parameters.
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 29]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
...
+    StatementFactory(Dialect dialect) {
+      this.dialect = dialect;
+    }
+
+    public Statement of(String sql) {
+      return Statement.of(sql);
+    }
+
+    /**
+     * This function accepts the SQL statement with unnamed parameters(?) and accepts the list of
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:292]:
```suggestion
     * This function accepts a SQL statement with unnamed parameters (?) and accepts a list of
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 30]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
...
+     * objects to replace unnamed parameters. Primitive types are supported.
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:293]:
```suggestion
     * objects that should be used as the values for those parameters. Primitive types are supported.
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 31]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
...
+     *
+     * <p>For Date column, following types are supported
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:295]:
```suggestion
     * <p>For parameters of type DATE, the following types are supported
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 32]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
...
+     *
+     * <ul>
+     *   <li>java.util.Date
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:298]:
```suggestion
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 33]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
...
+     *   <li>LocalDate
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:299]:
```suggestion
     *   <li>java.time.LocalDate
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 34]
```diff
@@ -246,4 +248,90 @@ StringBuilder toString(StringBuilder b) {
...
+     *   <li>com.google.cloud.Date
+     * </ul>
+     *
+     * <p>For Timestamp column, following types are supported. All the dates should be in UTC
+     * format. Incase if the timezone is not in UTC, spanner client will convert that to UTC
+     * automatically
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:305]:
```suggestion
     * <p>For parameters of type TIMESTAMP, the following types are supported. Note that Spanner stores all
     * timestamps in UTC. Instances of LocalDateTime and OffsetDateTime that use other timezones than UTC,
     * will be converted to the corresponding UTC values before being sent to Spanner.
     * Instances of LocalDateTime will be converted to a ZonedDateTime using the system default timezone,
     * and then converted to UTC before being sent to Spanner.
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java [hunk 35]
```diff
@@ -824,6 +838,156 @@ public static Value structArray(Type elementType, @Nullable Iterable<Struct> v)
 
   private Value() {}
 
+  static Value toValue(Object value) {
+    if (value instanceof Value) {
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:842]: We should add something like the following here:

```java
if (value == null) {
  return Value.untyped(null);
}
```

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ValueTest.java [hunk 36]
```diff
@@ -2496,6 +2508,302 @@ public void verifyBrokenSerialization() {
     reserializeAndAssert(BrokenSerializationList.of(1, 2, 3));
   }
 
+  @Test
+  public void testToValue() {
+    int i = 10;
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/ValueTest.java:2513]: Add a test for passing in a `null` to `Value.toValue(Object)`

**Review by olavloite — APPROVED** [review]
LGTM, with some minor nits on javadoc, and missing null handling

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java (hunk 35 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Value.java:845]: done

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ValueTest.java (hunk 36 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/ValueTest.java:2517]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java (hunk 27 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:269]: done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java [hunk 37]
```diff
@@ -246,4 +248,104 @@ StringBuilder toString(StringBuilder b) {
...
+   *
+   * SQL query with multiple parameters
+   *
+   * <pre>{@code
+   * long id = 10L;
+   * String name = "google";
+   * List<String> phoneNumbers = Arrays.asList("1234567890", "0987654321");
+   * Statement statement = databaseClient.getStatementFactory()
+   *    *     .withUnnamedParameters("INSERT INTO TABLE (ID, name, phonenumbers)
+   *       // VALUES(?, ?, ?)", id, name, phoneNumbers)
```
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java:278]: nit: this seems like a misformat / failed copy-paste

**Review by olavloite — APPROVED** [review]

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java (hunk 37 above)
Comment (sakthivelmanii) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Statement.java]: done
